plan/logical_plan_builder.go

```diff
@@ -266,6 +281,71 @@ func (b *planBuilder) buildJoin(join *ast.Join) LogicalPlan {
 	return joinPlan
 }
 
+func (b *planBuilder) buildUsingClause(p *LogicalJoin, leftPlan, rightPlan LogicalPlan, join *ast.Join) error {
+	lsc := leftPlan.Schema().Clone()
+	rsc := rightPlan.Schema().Clone()
+
+	schemaCols := make([]*expression.Column, 0, len(lsc.Columns)+len(rsc.Columns)-len(join.Using))
+	coalescedCols := make([]*expression.Column, 0, len(join.Using))
+	conds := make([]*expression.ScalarFunction, 0, len(join.Using))
+
+	coalesced := make(map[string]bool, len(join.Using))
+	for _, col := range join.Using {
+		var (
+			err    error
+			lc, rc *expression.Column
+			cond   expression.Expression
+		)
+
+		if lc, err = lsc.FindColumn(col); err != nil {
+			return errors.Trace(err)
+		}
+		if rc, err = rsc.FindColumn(col); err != nil {
+			return errors.Trace(err)
+		}
+		coalesced[col.Name.L] = true
+		if lc == nil || rc == nil {
+			return ErrUnknownColumn.GenByArgs(col.Name, "from clause")
+		}
+
+		if cond, err = expression.NewFunction(b.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), lc, rc); err != nil {
+			return errors.Trace(err)
+		}
+		conds = append(conds, cond.(*expression.ScalarFunction))
+		if join.Tp == ast.RightJoin {
+			schemaCols = append(schemaCols, rc)
+			coalescedCols = append(coalescedCols, lc)
+		} else {
+			schemaCols = append(schemaCols, lc)
+			coalescedCols = append(coalescedCols, rc)
+		}
+	}
+
+	sort.Slice(schemaCols, func(i, j int) bool {
+		return schemaCols[i].Position < schemaCols[j].Position
+	})
+
+	if join.Tp == ast.RightJoin {
+		lsc, rsc = rsc, lsc
+	}
+	for _, col := range lsc.Columns {
+		if !coalesced[col.ColName.L] {
+			schemaCols = append(schemaCols, col)
+		}
+	}
```

Comment:
According to MySQL 

> Redundant column elimination and column ordering occurs according to standard SQL, producing this display order:

> First, coalesced common columns of the two joined tables, in the order in which they occur in the first table

>Second, columns unique to the first table, in order in which they occur in that table

>Third, columns unique to the second table, in order in which they occur in that table

And the `first table` is the right hand side table in `right join`

So I first found the columns which in both table, then sort them based on their position in source table, then switch the `lsc` and `rsc` if there is a `right join`, finally I append remain columns in lsc and rsc.